Documenting another side-effect error:

https://productionresultssa9.blob.core.windows.net/actions-results/d8bfb864-5d97-47b7-a7d6-16ef0c4b7f86/workflow-job-run-179484e7-c0ae-5a16-24ca-1a74f27caf55/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-08-23T01%3A29%3A31Z&sig=huyfmqfgTWEZSwmoRUNtTiqvvUz4LV83y%2BNN8O6WFJY%3D&ske=2024-08-23T12%3A23%3A13Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-08-23T00%3A23%3A13Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-05-04&sp=r&spr=https&sr=b&st=2024-08-23T01%3A19%3A26Z&sv=2024-05-04

```
2024-08-22T20:35:00.0384777Z You have 18 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): galaxy, sessions, social_django.
2024-08-22T20:35:00.0386391Z Run 'python manage.py migrate' to apply them.
2024-08-22T20:35:00.0387182Z     resource = Resource.get_resource_for_object(instance)
2024-08-22T20:35:00.0387942Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0389658Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/models/resource.py", line 97, in get_resource_for_object
2024-08-22T20:35:00.0392396Z     return cls.objects.get(object_id=obj.pk, content_type=ContentType.objects.get_for_model(obj).pk)
2024-08-22T20:35:00.0393565Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0395035Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
2024-08-22T20:35:00.0396314Z     return getattr(self.get_queryset(), name)(*args, **kwargs)
2024-08-22T20:35:00.0397052Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0398151Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 637, in get
2024-08-22T20:35:00.0399219Z     raise self.model.DoesNotExist(
2024-08-22T20:35:00.0400358Z ansible_base.resource_registry.models.resource.Resource.DoesNotExist: Resource matching query does not exist.
2024-08-22T20:35:00.0401685Z 
2024-08-22T20:35:00.0402096Z During handling of the above exception, another exception occurred:
2024-08-22T20:35:00.0402747Z 
2024-08-22T20:35:00.0402934Z Traceback (most recent call last):
2024-08-22T20:35:00.0403815Z   File "/usr/local/bin/pulpcore-manager", line 8, in <module>
2024-08-22T20:35:00.0404528Z     sys.exit(manage())
2024-08-22T20:35:00.0404948Z              ^^^^^^^^
2024-08-22T20:35:00.0405930Z   File "/usr/local/lib/python3.11/site-packages/pulpcore/app/manage.py", line 11, in manage
2024-08-22T20:35:00.0406943Z     execute_from_command_line(sys.argv)
2024-08-22T20:35:00.0408359Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
2024-08-22T20:35:00.0409618Z     utility.execute()
2024-08-22T20:35:00.0411090Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
2024-08-22T20:35:00.0412396Z     self.fetch_command(subcommand).run_from_argv(self.argv)
2024-08-22T20:35:00.0413990Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 412, in run_from_argv
2024-08-22T20:35:00.0415217Z     self.execute(*args, **cmd_options)
2024-08-22T20:35:00.0416880Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/management/commands/createsuperuser.py", line 88, in execute
2024-08-22T20:35:00.0418366Z     return super().execute(*args, **options)
2024-08-22T20:35:00.0419008Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0420304Z   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 458, in execute
2024-08-22T20:35:00.0421480Z     output = self.handle(*args, **options)
2024-08-22T20:35:00.0422092Z              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0423682Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/management/commands/createsuperuser.py", line 233, in handle
2024-08-22T20:35:00.0425471Z     self.UserModel._default_manager.db_manager(database).create_superuser(
2024-08-22T20:35:00.0427173Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/models.py", line 172, in create_superuser
2024-08-22T20:35:00.0428567Z     return self._create_user(username, email, password, **extra_fields)
2024-08-22T20:35:00.0429497Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0430865Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/models.py", line 155, in _create_user
2024-08-22T20:35:00.0432376Z     user.save(using=self._db)
2024-08-22T20:35:00.0433570Z   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/base_user.py", line 76, in save
2024-08-22T20:35:00.0434613Z     super().save(*args, **kwargs)
2024-08-22T20:35:00.0435744Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/base.py", line 814, in save
2024-08-22T20:35:00.0436764Z     self.save_base(
2024-08-22T20:35:00.0437903Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/base.py", line 892, in save_base
2024-08-22T20:35:00.0438926Z     post_save.send(
2024-08-22T20:35:00.0440045Z   File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 176, in send
2024-08-22T20:35:00.0441087Z     return [
2024-08-22T20:35:00.0441468Z            ^
2024-08-22T20:35:00.0442669Z   File "/usr/local/lib/python3.11/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
2024-08-22T20:35:00.0443960Z     (receiver, receiver(signal=self, sender=sender, **named))
2024-08-22T20:35:00.0444715Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0446248Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/signals/handlers.py", line 34, in update_resource
2024-08-22T20:35:00.0447649Z     resource = init_resource_from_object(instance)
2024-08-22T20:35:00.0448412Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0450129Z   File "/usr/local/lib/python3.11/site-packages/ansible_base/resource_registry/models/resource.py", line 170, in init_resource_from_object
2024-08-22T20:35:00.0451910Z     resource_type = c_type.resource_type
2024-08-22T20:35:00.0452537Z                     ^^^^^^^^^^^^^^^^^^^^
2024-08-22T20:35:00.0454029Z   File "/usr/local/lib/python3.11/site-packages/django/db/models/fields/related_descriptors.py", line 492, in __get__
2024-08-22T20:35:00.0455415Z     raise self.RelatedObjectDoesNotExist(
2024-08-22T20:35:00.0456807Z django.contrib.contenttypes.models.ContentType.resource_type.RelatedObjectDoesNotExist: ContentType has no resource_type.
```

It looks like the error happens because resource_registry post-migrate logic runs _without ever having ran its own migrations_, which is a funny situation to be in.

--- a/ansible_base/lib/utils/db.py
+++ b/ansible_base/lib/utils/db.py
@@ -16,7 +16,7 @@ def ensure_transaction():
 
 
 def migrations_are_complete() -> bool:
-    """Returns a booling telling you if manage.py migrate has been run to completion
+    """Returns a boolean telling you if manage.py migrate has been run to completion
 
     Note that this is a little expensive, like up to 20 database queries
     and lots of imports.

--- a/ansible_base/lib/utils/db.py
+++ b/ansible_base/lib/utils/db.py
@@ -1,6 +1,7 @@
 from contextlib import contextmanager
 
-from django.db import transaction
+from django.db import connection, transaction
+from django.db.migrations.executor import MigrationExecutor
 
 
 @contextmanager
@@ -12,3 +13,15 @@ def ensure_transaction():
             yield
     else:
         yield
+
+
+def migrations_are_complete() -> bool:
+    """Returns a booling telling you if manage.py migrate has been run to completion
+
+    Note that this is a little expensive, like up to 20 database queries
+    and lots of imports.
+    Not suitable to run as part of a request, but expected to be okay
+    in a management command or post_migrate signals"""
+    executor = MigrationExecutor(connection)
+    plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
+    return not bool(plan)

--- a/ansible_base/resource_registry/apps.py
+++ b/ansible_base/resource_registry/apps.py
@@ -7,7 +7,7 @@ from django.db.models.functions import Cast
 from django.db.utils import IntegrityError
 
 import ansible_base.lib.checks  # noqa: F401 - register checks
-from ansible_base.lib.utils.db import ensure_transaction
+from ansible_base.lib.utils.db import ensure_transaction, migrations_are_complete
 
 logger = logging.getLogger('ansible_base.resource_registry.apps')
 
@@ -30,6 +30,10 @@ def initialize_resources(sender, **kwargs):
     # `django-admin flush` is called seems like a bad idea, since that will prevent the
     # resource types from being initialized in the database, so a direct import appears to be
     # better than doing nothing.
+
+    if not migrations_are_complete():
+        logger.info('Not running resource_registry post_migrate because migrations are incomplete')
+        return
 
     apps = kwargs.get("apps")
     if apps is None:

--- a/test_app/tests/lib/utils/test_db.py
+++ b/test_app/tests/lib/utils/test_db.py
@@ -1,0 +1,9 @@
+import pytest
+
+from ansible_base.lib.utils.db import migrations_are_complete
+
+
+@pytest.mark.django_db
+def test_migrations_are_complete():
+    "If you are running tests, migrations (test database) should be complete"
+    assert migrations_are_complete()

